sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -0,0 +1,164 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Internal shard iterators pool.
+ * It maintains the thread pool for reading Kinesis shards in separate threads.
+ * Read records are stored in a blocking queue of limited capacity.
+ */
+class ShardReadersPool {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ShardReadersPool.class);
+  private static final int DEFAULT_CAPACITY_PER_SHARD = 10_000;
+  private ExecutorService executorService;
+  private BlockingQueue<KinesisRecord> recordsQueue;
+  private Map<String, ShardRecordsIterator> shardIteratorsMap;
+  private SimplifiedKinesisClient kinesis;
+  private KinesisReaderCheckpoint initialCheckpoint;
+  private final int queueCapacityPerShard;
+  private AtomicBoolean poolOpened = new AtomicBoolean(true);
+
+  ShardReadersPool(SimplifiedKinesisClient kinesis, KinesisReaderCheckpoint initialCheckpoint) {
+    this(kinesis, initialCheckpoint, DEFAULT_CAPACITY_PER_SHARD);
+  }
+
+  ShardReadersPool(SimplifiedKinesisClient kinesis, KinesisReaderCheckpoint initialCheckpoint,
+      int queueCapacityPerShard) {
+    this.kinesis = kinesis;
+    this.initialCheckpoint = initialCheckpoint;
+    this.queueCapacityPerShard = queueCapacityPerShard;
+  }
+
+  void start() throws TransientKinesisException {
+    ImmutableMap.Builder<String, ShardRecordsIterator> shardsMap = ImmutableMap.builder();
+    for (ShardCheckpoint checkpoint : initialCheckpoint) {
+      shardsMap.put(checkpoint.getShardId(), createShardIterator(kinesis, checkpoint));
+    }
+    shardIteratorsMap = shardsMap.build();
+    executorService = Executors.newFixedThreadPool(shardIteratorsMap.size());
+    recordsQueue = new LinkedBlockingQueue<>(queueCapacityPerShard * shardIteratorsMap.size());
+    for (final ShardRecordsIterator shardRecordsIterator : shardIteratorsMap.values()) {
+      executorService.submit(new Runnable() {
+
+        @Override
+        public void run() {
+          readLoop(shardRecordsIterator);
+        }
+      });
+    }
+  }
+
+  private void readLoop(ShardRecordsIterator shardRecordsIterator) {
+    while (poolOpened.get()) {
+      try {
+        List<KinesisRecord> kinesisRecords = shardRecordsIterator.readNextBatch();
+        for (KinesisRecord kinesisRecord : kinesisRecords) {
+          recordsQueue.put(kinesisRecord);
+        }
+      } catch (TransientKinesisException e) {
+        LOG.warn("Transient exception occurred.", e);
+      } catch (InterruptedException e) {
+        LOG.warn("Thread was interrupted, finishing the read loop", e);
+        break;
+      } catch (Throwable e) {
+        LOG.error("Unexpected exception occurred", e);
+      }
+    }
+    LOG.info("Kinesis Shard read loop has finished");
+  }
+
+  CustomOptional<KinesisRecord> nextRecord() {
+    try {
+      KinesisRecord record = recordsQueue.poll(1, TimeUnit.SECONDS);
+      if (record == null) {
+        return CustomOptional.absent();
+      }
+      shardIteratorsMap.get(record.getShardId()).ackRecord(record);
```

Comment:
Ouch, this is not correct. Acking must happen in finalizeCheckpoint, not here - otherwise, if the bundle fails and gets retried, the record will be lost because it was already acked.

Reply:
As I replied in the other comment, there's no actual acking in Kinesis. So do you think this approach here is broken given that fact? I'm not that much familiar with Beam checkpointing.

Reply:
As I was speaking with Google guys some time ago, the order of operations in Beam seemed to follow the pattern:
* repeat many times:
** Reader.advance()
** Reader.getCurrent()
* Reader.getCheckpointMark()
and then checkpoint happens.

The important thing is that the checkpoint mark will be persisted by Beam only after the last row was successfully processed. This means that on each advance() we can safely create new CheckpointMark as @pawel-kaczmarczyk is doing here and simply give it to Beam framework once requested.